src/coreclr/jit/codegenarmarch.cpp

```diff
@@ -2094,34 +2675,202 @@ void CodeGen::genCodeForCpBlkUnroll(GenTreeBlk* node)
+        helper.TryEncodeAllOffsets(FP_REGSIZE_BYTES, &canEncodeAll16ByteWideLoads, &canEncodeAll16ByteWideStores);
+
+        if (canEncodeAll16ByteWideLoads && canEncodeAll16ByteWideStores)
         {
-            if (srcLclNum != BAD_VAR_NUM)
-            {
-                emit->emitIns_R_R_S_S(INS_ldp, EA_8BYTE, EA_8BYTE, tempReg, tempReg2, srcLclNum, srcOffset);
-            }
-            else
-            {
-                emit->emitIns_R_R_R_I(INS_ldp, EA_8BYTE, tempReg, tempReg2, srcAddrBaseReg, srcOffset);
-            }
+            // No further adjustments for srcOffset and dstOffset are needed.
+            // The JIT should use 16-byte loads and stores when the resulting sequence has fewer number of instructions.
 
-            if (dstLclNum != BAD_VAR_NUM)
-            {
-                emit->emitIns_S_S_R_R(INS_stp, EA_8BYTE, EA_8BYTE, tempReg, tempReg2, dstLclNum, dstOffset);
-            }
-            else
+            shouldUse16ByteWideInstrs =
+                (helper.InstructionCount(FP_REGSIZE_BYTES) < helper.InstructionCount(REGSIZE_BYTES));
+        }
+        else if (canEncodeAllLoads && canEncodeAllStores &&
+                 (canEncodeAll16ByteWideLoads || canEncodeAll16ByteWideStores))
+        {
+            // In order to use 16-byte instructions the JIT needs to adjust either srcOffset or dstOffset.
+            // The JIT should use 16-byte loads and stores when the resulting sequence (incl. an additional add
+            // instruction)
+            // has fewer number of instructions.
+
+            if (helper.InstructionCount(FP_REGSIZE_BYTES) + 1 < helper.InstructionCount(REGSIZE_BYTES))
             {
-                emit->emitIns_R_R_R_I(INS_stp, EA_8BYTE, tempReg, tempReg2, dstAddrBaseReg, dstOffset);
+                shouldUse16ByteWideInstrs = true;
+
+                if (!canEncodeAll16ByteWideLoads)
+                {
+                    srcOffsetAdj = srcOffset - dstOffset;
+                }
+                else
+                {
+                    dstOffsetAdj = dstOffset - srcOffset;
+                }
+
+                helper.SetSrcOffset(srcOffset - srcOffsetAdj);
+                helper.SetDstOffset(dstOffset - dstOffsetAdj);
             }
         }
     }
+
+#ifdef DEBUG
+    if (shouldUse16ByteWideInstrs)
+    {
+        assert(helper.CanEncodeAllOffsets(FP_REGSIZE_BYTES));
+    }
+    else
+    {
+        assert(helper.CanEncodeAllOffsets(REGSIZE_BYTES));
+    }
 #endif
 
+    regNumber intReg1 = REG_NA;
+    regNumber intReg2 = REG_NA;
+
+    // On Arm64 an internal integer register is always allocated for a CopyBlock node
+    // in addition to a reserved register.
+    // However, if it is determined that both srcOffset and dstOffset need adjustments
+    // than both registers will be used for storing the adjusted addresses.
+    // In that case, the JIT uses SIMD register(s) and instructions for copying.
+
+    const regNumber tempReg1 = rsGetRsvdReg();
```

Comment:
I guess the `rsGetRsvdReg()` might be available, and you know that the instructions and addressing modes you're generating won't require it (in the emitter), but it kind of seems wrong, anyway. It seems like you should request another temp reg from LSRA and use that instead.

One reason why: ideally, we would be able to only allocate `rsGetRsvdReg()` if the size of the frame (and maximum local variable offset) required it. We haven't done that on arm64 because we have "enough" registers, so always reserving one didn't seem expensive.